src/Psl/Type/Internal/VecType.php

```diff
@@ -55,23 +55,27 @@ public function matches(mixed $value): bool
     public function coerce(mixed $value): iterable
     {
         if (! is_iterable($value)) {
-            throw CoercionException::withValue($value, $this->toString(), $this->getTrace());
+            throw CoercionException::withValue($value, $this->toString());
         }
 
-        /** @var Type\Type<Tv> $value_type */
-        $value_type = $this->value_type->withTrace(
-            $this->getTrace()
-                ->withFrame($this->toString())
-        );
-
         /**
          * @var list<Tv> $entries
          */
         $result = [];
+        $value_type = $this->value_type;
 
-        /** @var Tv $v */
-        foreach ($value as $v) {
-            $result[] = $value_type->coerce($v);
+        /**
+         * @var Tv $v
+         * @var array-key $i
+         */
+        foreach ($value as $i => $v) {
+            try {
+                $v_result = $value_type->coerce($v);
+            } catch (CoercionException $e) {
+                throw CoercionException::withValue($v, $this->toString(), (string) $i, $e);
+            }
```

Comment:
A single `try` should indeed be a bit faster in the PHP engine, although `$i` and `$v` will be flagged as potentially unassigned by static analysis :thinking:

Reply:
much better indeed! (see new included benchmarks)